### PR TITLE
Add failing test that round trips a flattened signed amount in an option

### PR DESCRIPTION
**Pull request by casey**
Add a failing test that demonstrates a limitation using SignedAmount whose value is None in a flattened struct. I kind of suspect that this is an inherent limitation of `serde`, but I don't know enough to be sure.

Closing #266, which shows the issue in rust-bitcoincore-rpc-json, in favor of this issue.

### CODE REVIEW

**Comment by tcharding** [comment]
Is the link in the PR description correct?

**Comment by tcharding** [comment]
Thanks for the bug report @casey! It looks like the bug is ours, if we replace the `Option<SignedAmount>` with `Option<usize>` in `Bar` the test passes. Also the `visit_none` method we implement is never called. I'll dig some more.

**Comment by casey** [comment]
> Is the link in the PR description correct?

Ah, no sorry, I meant https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/266.

**Comment by tcharding** [comment]
I'm of the not-that-knowledgeable-about-serde opion that this may be a bug in serde. Our custom deserializer is called when using `flatten` and also the `visit_none` method is called when not using `flatten` but with `flatten` the `visit_none` method is never called.

**Comment by tcharding** [comment]
Oh? I just reached for the first inbuilt type that came to mind. Will try as you say using `i64`.

**Comment by tcharding** [comment]
Confirming using `Option<i64>` (also `f64`) instead of `Option<SignedAmount>` makes the test pass.

File: bitcoin/src/amount.rs [hunk 1]
```diff
@@ -2268,4 +2268,30 @@ mod tests {
...
+        struct Foo {
+            #[serde(flatten)]
+            bar: Bar,
+        }
+
+        #[derive(Serialize, Deserialize, Debug, PartialEq)]
+        #[serde(crate = "actual_serde")]
+        struct Bar {
+            #[serde(with = "crate::amount::serde::as_btc::opt")]
+            amount: Option<SignedAmount>,
```
Comment (tcharding) [inline bitcoin/src/amount.rs:2286]: If this PR ever gets rebased perhaps use `Amount` instead of `SignedAmount`, just because it saves devs wondering if this has anything to do with the signed-ness - which it doesn't since we use the same `opt` module for both.

Comment (Kixunil) [reply bitcoin/src/amount.rs:2286]: Ideally have both tests. I thought it has something to do with signedness and potentially `isize` vs `i64`. I suggested `i64` because that matches the inteded behavior of `SignedAmount` most closely.

I don't have much time for this now but my suggestion is to look into the implementation of `i64` and `u64` and compare the code.

Comment (tcharding) [reply bitcoin/src/amount.rs:2286]: Our implementation of `visit_none` is not being called when the `flatten` attribute is present but it is called when the attribute is not present - that is why I think its a bug in serde. And to verify it should work I did the "replace our type with a type from the stdlib".

**Comment by casey** [comment]
Cleaning up my stalled PRs. If someone wants this, feel free to open a new PR with the code.
